source/javascripts/controllers/_MainController.js.erb

```diff
@@ -604,141 +592,150 @@ const enableConfigConflictResolution = useFeatureFlag('enable-wfe-config-conflic
-          viewModel.deferredUserSavedYmlToRepository = $q.defer();
-        } else {
-          viewModel.deferredConflictResolution = $q.defer();
-        }
+        viewModel.saveProgress.start("<%= data[:strings][:main][:load_progress][:saving] %>");
 
-        var promise = $q(function (resolve, reject) {
-          if (viewModel.currentMenu.id === "secrets") {
-            return resolve();
+        function validateConfig() {
+          if (viewModel.currentMenu.id === "secrets" || viewModel.currentMenu.id === "yml") {
+            return $q.resolve();
           }
 
-          var reload = () => $q.all([
-            appService.getAppConfig(true),
-            appService.getAppConfigYML(true)
-          ]);
-          var saveAppConfig = appService.saveAppConfig;
-          var validateAppConfig = appService.validateAppConfig;
+          return appService.validateAppConfig();
+        }
+
+        function mergeChanges(error) {
+          var promise = $q.resolve();
           var finalizeAppConfig = appService.finalizeAppConfig;
-          var detectRemoteConfigChanges = appService.detectRemoteAppConfigChanges;
-          viewModel.dataToSave = appService.appConfig;
+          var detectRemoteAppConfigChanges = appService.detectRemoteAppConfigChanges;
+
+          viewModel.deferredConflictResolution = $q.defer();
+
+          if (!enableConfigConflictResolution || isRepositorySavedYml) {
+            return promise;
+          }
 
           if (viewModel.currentMenu.id === "yml") {
-            saveAppConfig = appService.saveAppConfigYML;
-            validateAppConfig = $q.resolve;
             finalizeAppConfig = appService.finalizeAppConfigYML;
-            detectRemoteConfigChanges = appService.detectRemoteAppConfigYMLChanges;
-            viewModel.dataToSave = appService.appConfigYML;
+            detectRemoteAppConfigChanges = appService.detectRemoteAppConfigYMLChanges;
           }
 
-          if (viewModel.currentMenu.id === "stack") {
-            reload = reloadStack;
-            saveAppConfig = appService.saveStackAndDockerImage;
+          if (error) {
+            viewModel.configMergeError = error.message;
+            promise = viewModel.deferredConflictResolution.promise;
+          } else {
+            promise = detectRemoteAppConfigChanges().then(function ({ hasChanges, yourYaml, baseYaml, remoteYaml }) {
+              if (!hasChanges) {
+                return viewModel.deferredConflictResolution.resolve();
+              }
+
+              viewModel.yourYaml = yourYaml;
+              viewModel.baseYaml = baseYaml;
+              viewModel.remoteYaml = remoteYaml;
+              viewModel.isConfigMergeDialogOpen = true;
+
+              return viewModel.deferredConflictResolution.promise;
+            })
           }
 
+          return promise.then((finalYaml) => {
+            if (!finalYaml) {
+              return;
+            }
+
+            viewModel.finalYaml = finalYaml;
```

Comment:
Still not convienced that we need the `viewModel.finalYaml`, since the `finalizeAppConfig` will write the `finalYaml` to `appService.appConfig` which in the next `detectRemoteAppConfigChanges` cycle will be used as `yourYaml`